fish-rust/src/wutil/wrealpath.rs

```diff
@@ -0,0 +1,70 @@
+use std::{ffi::OsStr, fs::canonicalize, os::unix::prelude::OsStrExt};
+
+use cxx::let_cxx_string;
+
+use crate::{
+    ffi::{str2wcstring, wcs2string},
+    wchar::{wstr, WString},
+    wchar_ffi::{WCharFromFFI, WCharToFFI},
+};
+
+/// Wide character realpath. The last path component does not need to be valid. If an error occurs,
+/// `wrealpath()` returns `None`
+pub fn wrealpath(pathname: &wstr) -> Option<WString> {
+    if pathname.is_empty() {
+        return None;
+    }
+
+    let mut real_path = Vec::new();
+    let mut narrow_path: Vec<u8> = wcs2string(&pathname.to_ffi()).from_ffi();
+
+    // Strip trailing slashes. This is treats "/a//" as equivalent to "/a" if /a is a non-directory.
+    while narrow_path.len() > 1 && narrow_path[narrow_path.len() - 1] == b'/' {
+        narrow_path.pop();
+    }
+
+    // `from_bytes` is Unix specific but there isn't really any other way to do this
+    // since `libc::realpath` is also Unix specific. I also don't think we support Windows
+    // outside of WSL + Cygwin (which should be fairly Unix-like anyways)
+    let narrow_res = canonicalize(OsStr::from_bytes(&narrow_path));
+
+    if let Ok(result) = narrow_res {
+        real_path.extend_from_slice(result.as_os_str().as_bytes());
```

Comment:
Please use `.into_os_string().into_vec()` instead of `as_os_str().as_bytes()` to avoid re-allocating.
Also just assign directly to `real_path` instead of calling `extend_from_slice()` because that might end up allocating again (I'm not sure what std currently does if calling that on a zero-size buffer, but just assigning is clearer because it makes it clear we're not appending).

Reply:
I've made this change; originally I thought it'd be cheaper to return a slice. And yes, `extend_from_slice` does allocate again. I think this change combined with the change you suggested to return `real_path` from the `if/else` does really help in reducing the number of allocations. Thank you so much!